Foundation/src/ProcessRunner.cpp

```diff
@@ -0,0 +1,227 @@
+using Poco::Path;
+using Poco::Stopwatch;
+
+
+namespace Poco {
+
+
+ProcessRunner::ProcessRunner(const std::string& cmd,
+		const Args& args,
+		const std::string& pidFile,
+		int options,
+		int timeout,
+		bool startProcess,
+		const Args& pidArgFmt): _cmd(cmd),
+			_args(args),
+			_pid(INVALID_PID),
+			_pidFile(pidFile),
+			_options(options),
+			_timeout(timeout),
+			_pPH(nullptr),
+			_started(false),
+			_logger(Poco::Logger::get("ProcessRunner"))
+{
+	if (!File(_cmd).exists())
+		throw Poco::NotFoundException(_cmd);
+
+	if (_pidFile.empty() && !_args.empty() && !pidArgFmt.empty())
+	{
+		// try to detect PID file name from arguments;
+		// we follow Poco convention
+		// (-pfile or --pidfile=file)
+		// TODO: windows
+		for (const auto& fmt : pidArgFmt)
+		{
+			for (const auto& arg : _args)
+			{
+				std::string a = Poco::trim(arg);
+				std::size_t pos = a.find(fmt);
+				if (pos == 0)
+				{
+					_pidFile = a.substr(fmt.length());
+					PIDFile::getFileName(_pidFile);
+					break;
+				}
+			}
+		}
+	}
+	if (startProcess) start();
+}
+
+
+ProcessRunner::~ProcessRunner()
+{
+	stop();
+	_t.join();
+}
+
+
+std::string ProcessRunner::cmdLine() const
+{
+	std::string cmdL = _cmd + ' ';
+	auto it = _args.begin();
+	auto end = _args.end();
+	for (; it != end;)
+	{
+		cmdL.append(*it);
+		if (++it == end) break;
+		cmdL.append(1, ' ');
+	}
+	return cmdL;
+}
+
+
+void ProcessRunner::run()
+{
+	ProcessHandle* pPH = nullptr;
+	int rc = 0;
+
+	try
+	{
+		_logger.information("Starting '%s', options=%d)", cmdLine(), _options);
+		_pPH.store(new ProcessHandle(Process::launch(_cmd, _args, _options)));
+		pPH = _pPH;
+		_pid = pPH->id();
+		rc = pPH->wait();
+	}
+	catch(const Poco::Exception& e)
+	{
+		_logger.error(Poco::format("run(): %s", e.displayText()));
+	}
+	catch(const std::exception& e)
+	{
+		_logger.error(Poco::format("run(): %s", std::string(e.what())));
+	}
+	catch(...)
+	{
+		_logger.error("run(): unknown error");
+	}
+
+	_pid = INVALID_PID;
+	_pPH.store(nullptr);
+
+	delete pPH;
+
+	if (rc) _logger.warning("process returned %d (0 expected)", rc);
+}
+
+
+void ProcessRunner::stop()
+{
+	if (_started)
+	{
+		// TODO:
+		// on Windows, this only works cleanly for Poco::Util::ServerApplication:
+		// https://github.com/pocoproject/poco/blob/ed43543113c5307e216615d8c004f5ffd8a8b37f/Foundation/include/Poco/Process.h#L252
+		// the solution is either to implement Windows SIGTERM handling in Poco:
+		// https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/signal?view=msvc-170
+		// or resort to brute force TerminateProcess Win API for non-Poco processes:
+		// https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess
+		ProcessHandle* pPH = nullptr;
+		if ((pPH = _pPH.exchange(nullptr)))
+			Process::requestTermination(pPH->id());
+
+		std::string msg = "Waiting for process shutdown";
+		if (!_pidFile.empty())
+		{
+			Poco::format(msg, "%s (pidFile: '%s')", msg, _pidFile);
```

Comment:
this one should first set msg to "Waiting for PID file", then format it with the name

Reply:
Fixed in [ca79ba8](https://github.com/pocoproject/poco/pull/4225/commits/ca79ba8cb058dba942dd9b86e912037325339d6e)